{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notebook for Joanito-2022 data processing \n",
    "### Developed by: Anna Maguza\n",
    "\n",
    "### Institute of Computational Biology - Computational Health Centre - Hemlholtz Munich\n",
    "\n",
    "### 18 October 2022"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as an\n",
    "import h5py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Setup Cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.6.0 anndata==0.7.4 umap==0.3.10 numpy==1.18.2 scipy==1.4.1 pandas==1.0.4 scikit-learn==0.22.2.post1 statsmodels==0.13.5 python-igraph==0.10.2 louvain==0.7.0 leidenalg==0.8.0\n"
     ]
    }
   ],
   "source": [
    "sc.settings.verbosity = 3             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Upload Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Data Upload (csv)\n",
    "Epithelial_meta = pd.read_csv('/lustre/groups/talaveralopez/datasets/Colorectal_cancer/Raw_data/Joanito, 2022/Raw Data/Epithelial_metadata.csv', index_col=[1])\n",
    "Nonepithelial_meta = pd.read_csv('/lustre/groups/talaveralopez/datasets/Colorectal_cancer/Raw_data/Joanito, 2022/Raw Data/NonEpithelial_metadata.csv', index_col=[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Data Upload (csv)\n",
    "Patientmeta = pd.read_csv('/lustre/groups/talaveralopez/datasets/Colorectal_cancer/Raw_data/Joanito, 2022/Raw Data/patient_clinical_information.csv', index_col=[1], encoding='cp1252')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For anndata==0.8.0 and higher"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_v3_10x_h5_mod(filename, *, start=None):\n",
    "    \"\"\"\n",
    "    Read hdf5 file from Cell Ranger v3 or later versions.\n",
    "    \"\"\"\n",
    "    with h5py.File(str(filename), 'r') as f:\n",
    "        try:\n",
    "            dsets = {}\n",
    "            sc.readwrite._collect_datasets(dsets, f[\"matrix\"])\n",
    "            from scipy.sparse import csr_matrix\n",
    "\n",
    "            M, N = dsets['shape']\n",
    "            data = dsets['data']\n",
    "            if dsets['data'].dtype == np.dtype('int32'):\n",
    "                data = dsets['data'].view('float32')\n",
    "                data[:] = dsets['data']\n",
    "            matrix = csr_matrix(\n",
    "                (data, dsets['indices'], dsets['indptr']),\n",
    "                shape=(N, M),\n",
    "            )\n",
    "            adata = an.AnnData(\n",
    "                matrix,\n",
    "                obs=dict(obs_names=dsets['barcodes'].astype(str)),\n",
    "                var=dict(\n",
    "                    var_names=dsets['name'].astype(str),\n",
    "                    feature_types=dsets['feature_type'].astype(str),\n",
    "                    genome=dsets['genome'].astype(str),\n",
    "                ),\n",
    "            )\n",
    "            return adata\n",
    "        except KeyError:\n",
    "            raise Exception('File is missing one or more required datasets.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For scanpy==1.6.0 and anndata==0.7.4 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _collect_datasets(dsets: dict, group: h5py.Group):\n",
    "    for k, v in group.items():\n",
    "        if isinstance(v, h5py.Dataset):\n",
    "            dsets[k] = v[()]\n",
    "        else:\n",
    "            _collect_datasets(dsets, v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_v3_10x_h5_mod2(filename, *, start=None):\n",
    "    \"\"\"\n",
    "    Read hdf5 file from Cell Ranger v3 or later versions.\n",
    "    \"\"\"\n",
    "    with h5py.File(str(filename), 'r') as f:\n",
    "        try:\n",
    "            dsets = {}\n",
    "            _collect_datasets(dsets, f[\"matrix\"])\n",
    "\n",
    "            from scipy.sparse import csr_matrix\n",
    "\n",
    "            M, N = dsets['shape']\n",
    "            data = dsets['data']\n",
    "            if dsets['data'].dtype == np.dtype('int32'):\n",
    "                data = dsets['data'].view('float32')\n",
    "                data[:] = dsets['data']\n",
    "            matrix = csr_matrix(\n",
    "                (data, dsets['indices'], dsets['indptr']),\n",
    "                shape=(N, M),\n",
    "            )\n",
    "            adata = an.AnnData(\n",
    "                matrix,\n",
    "                obs=dict(obs_names=dsets['barcodes'].astype(str)),\n",
    "                var=dict(\n",
    "                    var_names=dsets['name'].astype(str),\n",
    "                    feature_types=dsets['feature_type'].astype(str),\n",
    "                    genome=dsets['genome'].astype(str),\n",
    "                ),\n",
    "            )\n",
    "            return adata\n",
    "        except KeyError:\n",
    "            raise Exception('File is missing one or more required datasets.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Please use read_v3_10x_h5_mod2 for earlier scanpy and anndata versions and read_v3_10x_h5_mod for anndata later than 0.8.0\n",
    "Epithelian_data = read_v3_10x_h5_mod2('/lustre/groups/talaveralopez/datasets/Colorectal_cancer/Raw_data/Joanito, 2022/Raw Data/Epithelial_Count_matrix.h5')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Please use read_v3_10x_h5_mod2 for earlier scanpy and anndata versions and read_v3_10x_h5_mod for anndata later than 0.8.0\n",
    "Nonepithelian_data = read_v3_10x_h5_mod2('/lustre/groups/talaveralopez/datasets/Colorectal_cancer/Raw_data/Joanito, 2022/Raw Data/NonEpithelial_Count_matrix.h5')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Preparing anndata file (obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Rename Nonepithelial_meta\n",
    "Nonepithelial_meta = Nonepithelial_meta.rename(columns={\"nCount_RNA\": \"cell.ID\", \"msi\": \"MSS/MSI\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Nonepithelial_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Merging cells and patients\n",
    "merged_data_final = pd.concat([Epithelial_meta, Nonepithelial_meta])\n",
    "merged_data_final = pd.merge(merged_data_final, Patientmeta, \n",
    "                   on='patient.ID', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "del Epithelial_meta, Nonepithelial_meta, Patientmeta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                cell.ID  nFeature_RNA  percent.mt sample.ID  \\\n",
      "49150    KUL5_EXT129_CTGTGCTTCGCTTGTC-1          2926   15.401077    EXT129   \n",
      "49151    KUL5_EXT129_GAAATGAGTTCCTCCA-1          5779   13.561872    EXT129   \n",
      "49152    KUL5_EXT129_GATCGCGTCTGCTGTC-1          3934   19.530070    EXT129   \n",
      "49153    KUL5_EXT129_GGAACTTCAGGAATGC-1          5551   11.547377    EXT129   \n",
      "49154    KUL5_EXT129_TCACAAGGTTTGGGCC-1          3276    9.897333    EXT129   \n",
      "49155  CRC16_MUX8563_AAACCTGCAAGCCGCT-1          2577    1.235932   MUX8563   \n",
      "49156  CRC16_MUX8563_AAACCTGTCTCGATGA-1          1531    6.171890   MUX8563   \n",
      "49157  CRC16_MUX8563_AAACCTGTCTCTGTCG-1          3083    0.891850   MUX8563   \n",
      "49158  CRC16_MUX8563_AAACGGGCAAGTTAAG-1          1686    3.639915   MUX8563   \n",
      "49159  CRC16_MUX8563_AAACGGGCAGTTCATG-1          4304    3.975425   MUX8563   \n",
      "49160  CRC16_MUX8563_AAAGATGAGTTAACGA-1          3080    6.754530   MUX8563   \n",
      "\n",
      "      patient.ID sample.origin dataset_x   cell.type    iCMS    msi  ...  \\\n",
      "49150      SC044        Normal      KUL5  Epithelial  Normal  MSI-H  ...   \n",
      "49151      SC044        Normal      KUL5  Epithelial  Normal  MSI-H  ...   \n",
      "49152      SC044        Normal      KUL5  Epithelial  Normal  MSI-H  ...   \n",
      "49153      SC044        Normal      KUL5  Epithelial  Normal  MSI-H  ...   \n",
      "49154      SC044        Normal      KUL5  Epithelial  Normal  MSI-H  ...   \n",
      "49155    CRC2794         Tumor   CRC-SG1     PlasmaB     NaN    NaN  ...   \n",
      "49156    CRC2794         Tumor   CRC-SG1        T_NK     NaN    NaN  ...   \n",
      "49157    CRC2794         Tumor   CRC-SG1     PlasmaB     NaN    NaN  ...   \n",
      "49158    CRC2794         Tumor   CRC-SG1        T_NK     NaN    NaN  ...   \n",
      "49159    CRC2794         Tumor   CRC-SG1  Fibroblast     NaN    NaN  ...   \n",
      "49160    CRC2794         Tumor   CRC-SG1  Fibroblast     NaN    NaN  ...   \n",
      "\n",
      "      iCMS.transcriptomic iCMS.inferCNV  KRAS BRAF TP53  APC PIK3CA LymphNode  \\\n",
      "49150               iCMS3         iCMS3    wt   wt   wt   wt    NaN       0.0   \n",
      "49151               iCMS3         iCMS3    wt   wt   wt   wt    NaN       0.0   \n",
      "49152               iCMS3         iCMS3    wt   wt   wt   wt    NaN       0.0   \n",
      "49153               iCMS3         iCMS3    wt   wt   wt   wt    NaN       0.0   \n",
      "49154               iCMS3         iCMS3    wt   wt   wt   wt    NaN       0.0   \n",
      "49155               iCMS2         iCMS2    wt   wt  mut  mut     wt       0.0   \n",
      "49156               iCMS2         iCMS2    wt   wt  mut  mut     wt       0.0   \n",
      "49157               iCMS2         iCMS2    wt   wt  mut  mut     wt       0.0   \n",
      "49158               iCMS2         iCMS2    wt   wt  mut  mut     wt       0.0   \n",
      "49159               iCMS2         iCMS2    wt   wt  mut  mut     wt       0.0   \n",
      "49160               iCMS2         iCMS2    wt   wt  mut  mut     wt       0.0   \n",
      "\n",
      "      Normal Tumor  \n",
      "49150    2.0   3.0  \n",
      "49151    2.0   3.0  \n",
      "49152    2.0   3.0  \n",
      "49153    2.0   3.0  \n",
      "49154    2.0   3.0  \n",
      "49155    1.0   4.0  \n",
      "49156    1.0   4.0  \n",
      "49157    1.0   4.0  \n",
      "49158    1.0   4.0  \n",
      "49159    1.0   4.0  \n",
      "49160    1.0   4.0  \n",
      "\n",
      "[11 rows x 28 columns]\n"
     ]
    }
   ],
   "source": [
    "print(merged_data_final.loc[49150:49160, ])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Preparing anndata file (var and X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Concatenate 2 anndata files (nonepithelial and epithelian)\n",
    "Joanito_2022 = Nonepithelian_data.concatenate(Epithelian_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "del Nonepithelian_data, Epithelian_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Final Part in preparing anndata file: Merging obs (clinical and cell data) with X and var (gene counts and genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/icb/anna.maguza/miniconda3/envs/scnym_env/lib/python3.8/site-packages/anndata/_core/anndata.py:119: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    }
   ],
   "source": [
    "#Create anndata file \n",
    "Joanito_2022_anndata = an.AnnData(X=Joanito_2022.X,\n",
    "                        obs=merged_data_final,\n",
    "                        var=Joanito_2022.var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<373058x33287 sparse matrix of type '<class 'numpy.float32'>'\n",
       "\twith 741331835 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Chacking final anndata\n",
    "Joanito_2022_anndata.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38_ensembl93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38_ensembl93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38_ensembl93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38_ensembl93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38_ensembl93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC233755.2</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38_ensembl93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC233755.1</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38_ensembl93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC240274.1</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38_ensembl93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC213203.1</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38_ensembl93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM231C</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38_ensembl93</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33287 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               feature_types            genome\n",
       "MIR1302-2HG  Gene Expression  GRCh38_ensembl93\n",
       "FAM138A      Gene Expression  GRCh38_ensembl93\n",
       "OR4F5        Gene Expression  GRCh38_ensembl93\n",
       "AL627309.1   Gene Expression  GRCh38_ensembl93\n",
       "AL627309.3   Gene Expression  GRCh38_ensembl93\n",
       "...                      ...               ...\n",
       "AC233755.2   Gene Expression  GRCh38_ensembl93\n",
       "AC233755.1   Gene Expression  GRCh38_ensembl93\n",
       "AC240274.1   Gene Expression  GRCh38_ensembl93\n",
       "AC213203.1   Gene Expression  GRCh38_ensembl93\n",
       "FAM231C      Gene Expression  GRCh38_ensembl93\n",
       "\n",
       "[33287 rows x 2 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Chacking final anndata\n",
    "Joanito_2022_anndata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell.ID</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>sample.ID</th>\n",
       "      <th>patient.ID</th>\n",
       "      <th>sample.origin</th>\n",
       "      <th>dataset_x</th>\n",
       "      <th>cell.type</th>\n",
       "      <th>iCMS</th>\n",
       "      <th>msi</th>\n",
       "      <th>...</th>\n",
       "      <th>iCMS.transcriptomic</th>\n",
       "      <th>iCMS.inferCNV</th>\n",
       "      <th>KRAS</th>\n",
       "      <th>BRAF</th>\n",
       "      <th>TP53</th>\n",
       "      <th>APC</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th>LymphNode</th>\n",
       "      <th>Normal</th>\n",
       "      <th>Tumor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CRC16_MUX8563_AAACGGGGTCGATTGT-1</td>\n",
       "      <td>5099</td>\n",
       "      <td>17.969349</td>\n",
       "      <td>MUX8563</td>\n",
       "      <td>CRC2794</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>CRC-SG1</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>iCMS2</td>\n",
       "      <td>MSS</td>\n",
       "      <td>...</td>\n",
       "      <td>iCMS2</td>\n",
       "      <td>iCMS2</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>mut</td>\n",
       "      <td>mut</td>\n",
       "      <td>wt</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CRC16_MUX8563_AAAGATGCAGAAGCAC-1</td>\n",
       "      <td>4759</td>\n",
       "      <td>23.734351</td>\n",
       "      <td>MUX8563</td>\n",
       "      <td>CRC2794</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>CRC-SG1</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>iCMS2</td>\n",
       "      <td>MSS</td>\n",
       "      <td>...</td>\n",
       "      <td>iCMS2</td>\n",
       "      <td>iCMS2</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>mut</td>\n",
       "      <td>mut</td>\n",
       "      <td>wt</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CRC16_MUX8563_AAAGCAATCTAACGGT-1</td>\n",
       "      <td>2580</td>\n",
       "      <td>24.403016</td>\n",
       "      <td>MUX8563</td>\n",
       "      <td>CRC2794</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>CRC-SG1</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Normal</td>\n",
       "      <td>MSS</td>\n",
       "      <td>...</td>\n",
       "      <td>iCMS2</td>\n",
       "      <td>iCMS2</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>mut</td>\n",
       "      <td>mut</td>\n",
       "      <td>wt</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CRC16_MUX8563_ACAGCCGGTCTCTTAT-1</td>\n",
       "      <td>2499</td>\n",
       "      <td>11.020450</td>\n",
       "      <td>MUX8563</td>\n",
       "      <td>CRC2794</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>CRC-SG1</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>iCMS2</td>\n",
       "      <td>MSS</td>\n",
       "      <td>...</td>\n",
       "      <td>iCMS2</td>\n",
       "      <td>iCMS2</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>mut</td>\n",
       "      <td>mut</td>\n",
       "      <td>wt</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CRC16_MUX8563_ACAGCTATCCGTCATC-1</td>\n",
       "      <td>4937</td>\n",
       "      <td>23.243570</td>\n",
       "      <td>MUX8563</td>\n",
       "      <td>CRC2794</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>CRC-SG1</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>iCMS2</td>\n",
       "      <td>MSS</td>\n",
       "      <td>...</td>\n",
       "      <td>iCMS2</td>\n",
       "      <td>iCMS2</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>mut</td>\n",
       "      <td>mut</td>\n",
       "      <td>wt</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373053</th>\n",
       "      <td>KUL5_EXT129_TTTGTCAGTTGGACCC-1</td>\n",
       "      <td>1230</td>\n",
       "      <td>1.356960</td>\n",
       "      <td>EXT129</td>\n",
       "      <td>SC044</td>\n",
       "      <td>Normal</td>\n",
       "      <td>KUL5</td>\n",
       "      <td>PlasmaB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>iCMS3</td>\n",
       "      <td>iCMS3</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373054</th>\n",
       "      <td>KUL5_EXT129_TTTGTCATCAGCATGT-1</td>\n",
       "      <td>983</td>\n",
       "      <td>9.032534</td>\n",
       "      <td>EXT129</td>\n",
       "      <td>SC044</td>\n",
       "      <td>Normal</td>\n",
       "      <td>KUL5</td>\n",
       "      <td>T_NK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>iCMS3</td>\n",
       "      <td>iCMS3</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373055</th>\n",
       "      <td>KUL5_EXT129_TTTGTCATCCATGAGT-1</td>\n",
       "      <td>850</td>\n",
       "      <td>7.793522</td>\n",
       "      <td>EXT129</td>\n",
       "      <td>SC044</td>\n",
       "      <td>Normal</td>\n",
       "      <td>KUL5</td>\n",
       "      <td>T_NK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>iCMS3</td>\n",
       "      <td>iCMS3</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373056</th>\n",
       "      <td>KUL5_EXT129_TTTGTCATCGGTTCGG-1</td>\n",
       "      <td>1595</td>\n",
       "      <td>1.074276</td>\n",
       "      <td>EXT129</td>\n",
       "      <td>SC044</td>\n",
       "      <td>Normal</td>\n",
       "      <td>KUL5</td>\n",
       "      <td>PlasmaB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>iCMS3</td>\n",
       "      <td>iCMS3</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373057</th>\n",
       "      <td>KUL5_EXT129_TTTGTCATCTATCGCC-1</td>\n",
       "      <td>1532</td>\n",
       "      <td>1.153075</td>\n",
       "      <td>EXT129</td>\n",
       "      <td>SC044</td>\n",
       "      <td>Normal</td>\n",
       "      <td>KUL5</td>\n",
       "      <td>PlasmaB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>iCMS3</td>\n",
       "      <td>iCMS3</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>wt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>373058 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 cell.ID  nFeature_RNA  percent.mt sample.ID  \\\n",
       "0       CRC16_MUX8563_AAACGGGGTCGATTGT-1          5099   17.969349   MUX8563   \n",
       "1       CRC16_MUX8563_AAAGATGCAGAAGCAC-1          4759   23.734351   MUX8563   \n",
       "2       CRC16_MUX8563_AAAGCAATCTAACGGT-1          2580   24.403016   MUX8563   \n",
       "3       CRC16_MUX8563_ACAGCCGGTCTCTTAT-1          2499   11.020450   MUX8563   \n",
       "4       CRC16_MUX8563_ACAGCTATCCGTCATC-1          4937   23.243570   MUX8563   \n",
       "...                                  ...           ...         ...       ...   \n",
       "373053    KUL5_EXT129_TTTGTCAGTTGGACCC-1          1230    1.356960    EXT129   \n",
       "373054    KUL5_EXT129_TTTGTCATCAGCATGT-1           983    9.032534    EXT129   \n",
       "373055    KUL5_EXT129_TTTGTCATCCATGAGT-1           850    7.793522    EXT129   \n",
       "373056    KUL5_EXT129_TTTGTCATCGGTTCGG-1          1595    1.074276    EXT129   \n",
       "373057    KUL5_EXT129_TTTGTCATCTATCGCC-1          1532    1.153075    EXT129   \n",
       "\n",
       "       patient.ID sample.origin dataset_x   cell.type    iCMS  msi  ...  \\\n",
       "0         CRC2794         Tumor   CRC-SG1  Epithelial   iCMS2  MSS  ...   \n",
       "1         CRC2794         Tumor   CRC-SG1  Epithelial   iCMS2  MSS  ...   \n",
       "2         CRC2794         Tumor   CRC-SG1  Epithelial  Normal  MSS  ...   \n",
       "3         CRC2794         Tumor   CRC-SG1  Epithelial   iCMS2  MSS  ...   \n",
       "4         CRC2794         Tumor   CRC-SG1  Epithelial   iCMS2  MSS  ...   \n",
       "...           ...           ...       ...         ...     ...  ...  ...   \n",
       "373053      SC044        Normal      KUL5     PlasmaB     NaN  NaN  ...   \n",
       "373054      SC044        Normal      KUL5        T_NK     NaN  NaN  ...   \n",
       "373055      SC044        Normal      KUL5        T_NK     NaN  NaN  ...   \n",
       "373056      SC044        Normal      KUL5     PlasmaB     NaN  NaN  ...   \n",
       "373057      SC044        Normal      KUL5     PlasmaB     NaN  NaN  ...   \n",
       "\n",
       "       iCMS.transcriptomic iCMS.inferCNV  KRAS BRAF TP53  APC PIK3CA  \\\n",
       "0                    iCMS2         iCMS2    wt   wt  mut  mut     wt   \n",
       "1                    iCMS2         iCMS2    wt   wt  mut  mut     wt   \n",
       "2                    iCMS2         iCMS2    wt   wt  mut  mut     wt   \n",
       "3                    iCMS2         iCMS2    wt   wt  mut  mut     wt   \n",
       "4                    iCMS2         iCMS2    wt   wt  mut  mut     wt   \n",
       "...                    ...           ...   ...  ...  ...  ...    ...   \n",
       "373053               iCMS3         iCMS3    wt   wt   wt   wt    NaN   \n",
       "373054               iCMS3         iCMS3    wt   wt   wt   wt    NaN   \n",
       "373055               iCMS3         iCMS3    wt   wt   wt   wt    NaN   \n",
       "373056               iCMS3         iCMS3    wt   wt   wt   wt    NaN   \n",
       "373057               iCMS3         iCMS3    wt   wt   wt   wt    NaN   \n",
       "\n",
       "       LymphNode Normal Tumor  \n",
       "0            0.0    1.0   4.0  \n",
       "1            0.0    1.0   4.0  \n",
       "2            0.0    1.0   4.0  \n",
       "3            0.0    1.0   4.0  \n",
       "4            0.0    1.0   4.0  \n",
       "...          ...    ...   ...  \n",
       "373053       0.0    2.0   3.0  \n",
       "373054       0.0    2.0   3.0  \n",
       "373055       0.0    2.0   3.0  \n",
       "373056       0.0    2.0   3.0  \n",
       "373057       0.0    2.0   3.0  \n",
       "\n",
       "[373058 rows x 28 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Chacking final anndata\n",
    "Joanito_2022_anndata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'sample.ID' as categorical\n",
      "... storing 'patient.ID' as categorical\n",
      "... storing 'sample.origin' as categorical\n",
      "... storing 'dataset_x' as categorical\n",
      "... storing 'cell.type' as categorical\n",
      "... storing 'iCMS' as categorical\n",
      "... storing 'msi' as categorical\n",
      "... storing 'dataset_y' as categorical\n",
      "... storing 'Gender' as categorical\n",
      "... storing 'MSS/MSI' as categorical\n",
      "... storing 'Site' as categorical\n",
      "... storing 'Sidedness' as categorical\n",
      "... storing 'Group Stage' as categorical\n",
      "... storing 'Stage TNM' as categorical\n",
      "... storing 'iCMS.transcriptomic' as categorical\n",
      "... storing 'iCMS.inferCNV' as categorical\n",
      "... storing 'KRAS' as categorical\n",
      "... storing 'BRAF' as categorical\n",
      "... storing 'TP53' as categorical\n",
      "... storing 'APC' as categorical\n",
      "... storing 'PIK3CA' as categorical\n",
      "... storing 'feature_types' as categorical\n",
      "... storing 'genome' as categorical\n"
     ]
    }
   ],
   "source": [
    "#Saving the final anndata file\n",
    "Joanito_2022_anndata.write('/lustre/groups/talaveralopez/datasets/Colorectal_cancer/Raw_data/Joanito, 2022/Anndata Raw/joanito_2022_anndata_raw-2.h5ad')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "38196666d5885f2c7b712fecdbb810606921747bf937f5c5e660684fe79200fa"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
